CIPs/0005.md

```diff
@@ -0,0 +1,74 @@
+
+Issue https://github.com/celo-org/celo-monorepo/issues/2667 tracks this component.
+
+### Fast discovery of changed enodeURL mapping
+
+This component will require all full nodes and validators to store a copy of a table that will contain the latest enodeURL mapping version number for each elected validator.  Whenver a validator updates it's enodeURL mapping, it will need to gossip a message containing the updated version number, and all nodes should update their table.
+
+Whenever a connection is established between two nodes, they will synchronize their tables so that both tables will have the latest version numbers.  This will handle the case of when a node just joins the P2P network and when a network partition has been repaired.  In those cases, the nodes throughout the network should quickly synchronize their tables.
+
+Issue https://github.com/celo-org/celo-monorepo/issues/2668 tracks this component.
+
+### EnodeURL query message
+
+When a validator realizes that it's assigned enodeURL for a remote validator has been updated, that validator should gossip a enodeURL query message destined to the remote validator.  Once the destined validator receives the message, it will send a direct announce message to the originating validator.
+
+Note that the originating validator will need to continuously re-gossip out this message until it receives a direct announce message, since it will never know for sure if the destined validator got the message (e.g. a network may have become partitioned).  The period between regossips should initially be 5 minutes, and doubled after every subsequent regossip, and finally capped at 1 hour.
+
+Issue https://github.com/celo-org/celo-blockchain/issues/863 tracks this component.
+
+## Alternative Solutions
+
+There were two variants to this proposals that were considered.
+
+1) Not including the fast discovery of changed enodeURL mapping, and instead have validators send enodeURL query messages once they realize that they are not connected to an elected validator.  The proposal will not handle the case when a validator is down as elegantly.  In that case, all the other validators will continuously be gossiping query messages until it's back up, which could result in high bandwidth usage.
+
+2) Persisting the enodeURL query gossip message for all full nodes/validators and have them be regossiped on peer connection.  This will reduce any need to regossip query messages and I think can be considered to be added later.  This does add significant complexity to the proposal above and probably handles a case that rarely happens.  If we find out that it happens more then expected, then this feature could be added.
+
+## Risks
+
+These protocol changes has a direct impact on consensus since the IBFT overlay network depends on this protocol.
```

Comment:
I think we should be more clear here that this proposal is not interoperable with the existing announce protocol and this change could not be deployed to a live network as proposed.